src/worker/ingestion/action-matcher.ts

```diff
@@ -0,0 +1,498 @@
+            case 'event':
+                return this.checkEventAgainstEventFilter(event, filter)
+            case 'person':
+                return this.checkEventAgainstPersonFilter(event, person, filter)
+            case 'element':
+                return this.checkEventAgainstElementFilter(elements, filter)
+            case 'cohort':
+                return await this.checkEventAgainstCohortFilter(event, person, filter)
+            default:
+                return false
+        }
+    }
+
+    /**
+     * Sublevel 4 of action matching.
+     */
+    private checkEventAgainstEventFilter(event: PluginEvent, filter: EventPropertyFilter): boolean {
+        return this.checkPropertiesAgainstFilter(event.properties, filter)
+    }
+
+    /**
+     * Sublevel 4 of action matching.
+     */
+    private checkEventAgainstPersonFilter(
+        event: PluginEvent,
+        person: Person | undefined,
+        filter: PersonPropertyFilter
+    ): boolean {
+        if (!person?.properties) {
+            return false
+        }
+        return this.checkPropertiesAgainstFilter(person.properties, filter) // TODO: get person here for use instead of event
+    }
+
+    /**
+     * Sublevel 4 of action matching.
+     */
+    private checkEventAgainstElementFilter(elements: Element[], filter: ElementPropertyFilter): boolean {
+        if (filter.key === 'selector') {
+            return this.checkElementsAgainstSelector(elements, filter.value)
+        } else {
+            return elements.some((element) => this.checkPropertiesAgainstFilter(element, filter))
+        }
+    }
+
+    /**
+     * Sublevel 4 of action matching.
+     */
+    private async checkEventAgainstCohortFilter(
+        event: PluginEvent,
+        person: Person | undefined,
+        filter: CohortPropertyFilter
+    ): Promise<boolean> {
+        if (!person?.properties) {
+            return false
+        }
+        return await this.db.doesPersonBelongToCohort(parseInt((filter.value as unknown) as any), person.id)
+    }
+
+    /**
+     * Sublevel 5 of action matching.
+     */
+    private checkPropertiesAgainstFilter(
+        properties: Properties | null | undefined,
+        filter: PropertyFilterWithOperator
+    ): boolean {
+        if (!properties) {
+            return false // MISMATCH DUE TO LACK OF PROPERTIES THAT COULD FULFILL CONDITION
+        }
+
+        const possibleValues = Array.isArray(filter.value) ? filter.value : [filter.value]
+        const foundValue = properties[filter.key]
+        let foundValueLowerCase: string // only calculated if needed for a case-insensitive operator
+
+        const requiredValueType = propertyOperatorToRequiredValueType[filter.operator]
+        if (requiredValueType && typeof foundValue !== requiredValueType) {
+            return false // MISMATCH DUE TO VALUE TYPE INCOMPATIBLE WITH OPERATOR SUPPORT
+        }
+
+        let test: (possibleValue: any) => boolean
+        switch (filter.operator) {
```

Comment:
nit (personal preference?): this could be an object mapping `filter.operator` to `test` func?

Reply:
We couldn't do the optimization of only calculating `foundValueLowerCase` in case of case-insensitive operators. Well, we could, but it'd be uglier.